File: frontend/src/engine/moveValidation.ts
import { BOARD_SIZE } from './constants';
import type { GameState, PlayerIndex, Position, Wall } from './gameTypes';
import { isMovementBlocked, wallsEqual, wallsIntersect, wouldWallPostOverlap } from './wallUtils';

function isOnBoard(pos: Position): boolean {
  return pos.row >= 0 && pos.row < BOARD_SIZE && pos.col >= 0 && pos.col < BOARD_SIZE;
}

function positionsEqual(a: Position, b: Position): boolean {
  return a.row === b.row && a.col === b.col;
}

/**
 * Port of GameEngine.getValidMoves
 * Returns valid pawn positions for the given player.
 */
export function getValidPawnMoves(state: GameState, playerIndex: PlayerIndex): Position[] {
  const player = state.players[playerIndex];
  const opponent = state.players[playerIndex === 0 ? 1 : 0];
  const validMoves: Position[] = [];

  const directions: Position[] = [
    { row: -1, col: 0 },
    { row: 1, col: 0 },
    { row: 0, col: -1 },
    { row: 0, col: 1 },
  ];

  for (const dir of directions) {
    const newPos: Position = {
      row: player.position.row + dir.row,
      col: player.position.col + dir.col,
    };

    if (!isOnBoard(newPos)) continue;
    if (isMovementBlocked(player.position, newPos, state.walls)) continue;

    if (positionsEqual(newPos, opponent.position)) {
      // Try to jump straight over
      const jumpPos: Position = {
        row: newPos.row + dir.row,
        col: newPos.col + dir.col,
      };

      if (
        isOnBoard(jumpPos) &&
        !isMovementBlocked(newPos, jumpPos, state.walls) &&
        !positionsEqual(jumpPos, player.position)
      ) {
        validMoves.push(jumpPos);
      } else {
        // Straight jump blocked – try diagonal jumps
        const diagDirections: Position[] =
          dir.row !== 0
            ? [
                { row: 0, col: -1 },
                { row: 0, col: 1 },
              ]
            : [
                { row: -1, col: 0 },
                { row: 1, col: 0 },
              ];

        for (const diagDir of diagDirections) {
          const diagJumpPos: Position = {
            row: newPos.row + diagDir.row,
            col: newPos.col + diagDir.col,
          };

          if (
            isOnBoard(diagJumpPos) &&
            !isMovementBlocked(newPos, diagJumpPos, state.walls) &&
            !positionsEqual(diagJumpPos, player.position)
          ) {
            validMoves.push(diagJumpPos);
          }
        }
      }
    } else {
      validMoves.push(newPos);
    }
  }

  return validMoves;
}

/**
 * BFS to check if there is a path from startPos to goalRow without any player blocking.
 * Only considers wall connectivity.
 */
export function hasPathToGoal(
  startPos: Position,
  goalRow: number,
  walls: readonly Wall[],
): boolean {
  const visited = new Set<string>();
  const queue: Position[] = [startPos];
  visited.add(`${startPos.row},${startPos.col}`);

  const directions: Position[] = [
    { row: -1, col: 0 },
    { row: 1, col: 0 },
    { row: 0, col: -1 },
    { row: 0, col: 1 },
  ];

  while (queue.length > 0) {
    const current = queue.shift()!;
    if (current.row === goalRow) return true;

    for (const dir of directions) {
      const next: Position = { row: current.row + dir.row, col: current.col + dir.col };
      const key = `${next.row},${next.col}`;
      if (
        isOnBoard(next) &&
        !visited.has(key) &&
        !isMovementBlocked(current, next, walls)
      ) {
        visited.add(key);
        queue.push(next);
      }
    }
  }

  return false;
}

/**
 * Port of GameEngine.isValidFencePlacement
 */
export function isValidWallPlacement(state: GameState, wall: Wall): boolean {
  // Bounds check: row/col 0-7 (anchor spans 2 squares)
  if (wall.row < 0 || wall.row > 7 || wall.col < 0 || wall.col > 7) return false;

  // Duplicate check
  if (state.walls.some((w) => wallsEqual(w, wall))) return false;

  // Post overlap check
  if (wouldWallPostOverlap(wall, state.walls)) return false;

  // Intersection check
  for (const existing of state.walls) {
    if (wallsIntersect(wall, existing)) return false;
  }

  // Path connectivity check with the new wall added
  const newWalls = [...state.walls, wall];
  for (const player of state.players) {
    if (!hasPathToGoal(player.position, player.goalRow, newWalls)) return false;
  }

  return true;
}

File: backend/app/engine/move_validation.py
"""Move legality + path-to-goal — mirrors frontend/src/engine/moveValidation.ts."""

from __future__ import annotations

from collections import deque
from collections.abc import Iterable

from app.engine.constants import BOARD_SIZE
from app.engine.game_types import GameState, PlayerIndex, Position, Wall
from app.engine.wall_utils import (
    is_movement_blocked,
    walls_equal,
    walls_intersect,
    would_wall_post_overlap,
)

_DIRS = (
    Position(-1, 0),
    Position(1, 0),
    Position(0, -1),
    Position(0, 1),
)


def _on_board(p: Position) -> bool:
    return 0 <= p.row < BOARD_SIZE and 0 <= p.col < BOARD_SIZE


def get_valid_pawn_moves(state: GameState, player_index: PlayerIndex) -> list[Position]:
    player = state.players[player_index]
    opponent = state.players[1 - player_index]
    moves: list[Position] = []

    for d in _DIRS:
        new_pos = Position(player.position.row + d.row, player.position.col + d.col)
        if not _on_board(new_pos):
            continue
        if is_movement_blocked(player.position, new_pos, state.walls):
            continue

        if new_pos == opponent.position:
            jump = Position(new_pos.row + d.row, new_pos.col + d.col)
            if (
                _on_board(jump)
                and not is_movement_blocked(new_pos, jump, state.walls)
                and jump != player.position
            ):
                moves.append(jump)
            else:
                # Straight jump blocked — try diagonal jumps perpendicular to dir
                if d.row != 0:
                    diag_dirs = (Position(0, -1), Position(0, 1))
                else:
                    diag_dirs = (Position(-1, 0), Position(1, 0))
                for dd in diag_dirs:
                    diag = Position(new_pos.row + dd.row, new_pos.col + dd.col)
                    if (
                        _on_board(diag)
                        and not is_movement_blocked(new_pos, diag, state.walls)
                        and diag != player.position
                    ):
                        moves.append(diag)
        else:
            moves.append(new_pos)

    return moves


def has_path_to_goal(start: Position, goal_row: int, walls: Iterable[Wall]) -> bool:
    """BFS that respects walls only — opponent positions are not considered blockers."""
    walls_t = tuple(walls)
    visited: set[tuple[int, int]] = {(start.row, start.col)}
    queue: deque[Position] = deque([start])
    while queue:
        cur = queue.popleft()
        if cur.row == goal_row:
            return True
        for d in _DIRS:
            nxt = Position(cur.row + d.row, cur.col + d.col)
            key = (nxt.row, nxt.col)
            if key in visited or not _on_board(nxt):
                continue
            if is_movement_blocked(cur, nxt, walls_t):
                continue
            visited.add(key)
            queue.append(nxt)
    return False


def is_valid_wall_placement(state: GameState, wall: Wall) -> bool:
    if not (0 <= wall.row <= 7 and 0 <= wall.col <= 7):
        return False
    if any(walls_equal(w, wall) for w in state.walls):
        return False
    if would_wall_post_overlap(wall, state.walls):
        return False
    if any(walls_intersect(wall, existing) for existing in state.walls):
        return False
    new_walls = (*state.walls, wall)
    for player in state.players:
        if not has_path_to_goal(player.position, player.goal_row, new_walls):
            return False
    return True
